src/Gui/ViewProviderGroupExtension.cpp

```diff
@@ -171,24 +218,56 @@ void ViewProviderGroupExtension::extensionHide() {
+bool ViewProviderGroupExtension::extensionOnDelete(const std::vector< std::string >& subNames) {
 
     auto* group = getExtendedViewProvider()->getObject()->getExtensionByType<App::GroupExtension>();
-    // If the group is nonempty ask the user if they want to delete its content
-    if (group->Group.getSize() > 0) {
-        QMessageBox::StandardButton choice =
-            QMessageBox::question(getMainWindow(), QObject::tr ( "Delete group content?" ),
-                QObject::tr ( "The %1 is not empty, delete its content as well?")
-                    .arg ( QString::fromUtf8 ( getExtendedViewProvider()->getObject()->Label.getValue () ) ),
-                QMessageBox::Yes | QMessageBox::No, QMessageBox::Yes );
-
-        if (choice == QMessageBox::Yes) {
-            Gui::Command::doCommand(Gui::Command::Doc,
-                    "App.getDocument(\"%s\").getObject(\"%s\").removeObjectsFromDocument()"
-                    , getExtendedViewProvider()->getObject()->getDocument()->getName()
-                    , getExtendedViewProvider()->getObject()->getNameInDocument());
-        }
+    
+    std::vector<App::DocumentObject*> directChildren = group->Group.getValues();
+    
+    // just delete without messagebox if group is empty
+    if (directChildren.empty()) {
+        return true;
+    }
+    
+    std::vector<App::DocumentObject*> allDescendants;
+    if (getExtendedViewProvider()->getObject()->isDerivedFrom<App::DocumentObjectGroup>()) {
+        auto* docGroup = static_cast<App::DocumentObjectGroup*>(getExtendedViewProvider()->getObject());
+        allDescendants = docGroup->getAllChildren();
+    } else {
+        allDescendants = directChildren;
+    }
+    
+    QString message;
+    if (allDescendants.size() == directChildren.size()) {
+        message = QObject::tr("The group '%1' contains %2 object(s). Do you want to delete them as well?")
+                      .arg(QString::fromUtf8(getExtendedViewProvider()->getObject()->Label.getValue()))
+                      .arg(allDescendants.size());
+    } else {
+        // if we have nested groups
+        message = QObject::tr("The group '%1' contains %2 direct children and %3 total descendants (including nested groups). Do you want to delete all of them recursively?")
+                      .arg(QString::fromUtf8(getExtendedViewProvider()->getObject()->Label.getValue()))
+                      .arg(directChildren.size())
+                      .arg(allDescendants.size());
+    }
+    
+    QMessageBox::StandardButton choice = QMessageBox::question(
+        getMainWindow(),
+        QObject::tr("Delete group contents recursively?"),
+        message,
+        QMessageBox::Yes | QMessageBox::No | QMessageBox::Cancel,
+        QMessageBox::No
+    );
+    
+    if (choice == QMessageBox::Cancel) {
+        // don't delete anything if user has cancelled
+        return false;
+    }
+    else if (choice == QMessageBox::Yes) {
```

Comment:
do not use 'else' after 'return' [readability-else-after-return]
```suggestion
    if (choice == QMessageBox::Yes) {
```

Reply:
Done.